bin/phase-helpers.sh

```diff
@@ -1094,23 +1094,44 @@ if ___eapi_has_eapply_user; then
 
 		local basedir=${PORTAGE_CONFIGROOT%/}/etc/portage/patches
 
-		local d applied
+		local applied d f
+		local -A _eapply_user_patches
 		local prev_shopt=$(shopt -p nullglob)
 		shopt -s nullglob
 
-		# possibilities:
+		# Patches from all matched directories are combined into a
+		# sorted (POSIX order) list of the patch basenames. Patches
+		# in more-specific directories override patches of the same
+		# basename found in less-specific directories. An empty patch
+		# (or /dev/null symlink) negates a patch with the same
+		# basename found in a less-specific directory.
+		#
+		# order of specificity:
 		# 1. ${CATEGORY}/${P}-${PR} (note: -r0 desired to avoid applying
 		#    ${P} twice)
 		# 2. ${CATEGORY}/${P}
 		# 3. ${CATEGORY}/${PN}
 		# all of the above may be optionally followed by a slot
-		for d in "${basedir}"/${CATEGORY}/{${P}-${PR},${P},${PN}}{,:${SLOT%/*}}; do
-			if [[ -n $(echo "${d}"/*.diff) || -n $(echo "${d}"/*.patch) ]]; then
-				eapply "${d}"
-				applied=1
-			fi
+		for d in "${basedir}"/${CATEGORY}/{${P}-${PR},${P},${PN}}{:${SLOT%/*},}; do
+			for f in "${d}"/*; do
+				if [[ ( ${f} == *.diff || ${f} == *.patch ) &&
+					-z ${_eapply_user_patches[${f##*/}]} ]]; then
+					_eapply_user_patches[${f##*/}]=${f}
+				fi
+			done
 		done
 
+		if [[ ${#_eapply_user_patches[@]} -gt 0 ]]; then
+			while read -r -d '' f; do
+				f=${_eapply_user_patches[${f}]}
+				if [[ -s ${f} ]]; then
+					eapply "${f}"
+					applied=1
+				fi
+			done < <(printf -- "%s\0" "${!_eapply_user_patches[@]}" |
```

Comment:
shellcheck:  ^-- SC1117: Backslash is literal in "\0". Prefer explicit escaping: "\\0".
https://github.com/koalaman/shellcheck/wiki/SC1117